### PR TITLE
feat: Migrate `LogSchema` `source_type_key` to new lookup code

### PR DESCRIPTION
**Pull request by pront**
This part of https://github.com/vectordotdev/vector/issues/13033.

Tried this with a config where `source_type_key` is:
* `""` --> no source entry
* not defined --> same as above
* `"foo"` --> `"foo":"demo_logs"`
* `"foo.bar"` --> `"foo":{"bar":"demo_logs"}`
* `"foo.."` --> `Configuration error. error=Invalid field path "foo.."`

The config:
```
data_dir = "/Users/pavlos.rontidis/my_tests/vector"

[log_schema]
source_type_key = "foo"

[sources.source0]
format = "json"
type = "demo_logs"

[sinks.sink0]
inputs = ["source0"]
target = "stdout"
type = "console"

[sinks.sink0.encoding]
codec = "json"
```

### CODE REVIEW

File: src/sources/datadog_agent/mod.rs [hunk 1]
```diff
@@ -334,7 +334,8 @@ impl DatadogAgentSource {
                     .expect("static regex always compiles"),
             },
             log_schema_host_key: log_schema().host_key(),
-            log_schema_source_type_key: log_schema().source_type_key(),
+            log_schema_source_type_key: log_schema().source_type_key()
+                .map_or("", |key| Box::leak(key.to_string().into_boxed_str())),
```
Comment (pront) [inline src/sources/datadog_agent/mod.rs:338]: An alternative here is to change `log_schema_source_type_key` which is `&'static str` to `String`. That would also require changing a couple of [places](https://github.com/search?q=repo%3Avectordotdev%2Fvector%20%22source.log_schema_source_type_key%2C%22&type=code) to copy.

File: lib/vector-core/src/event/log_event.rs [hunk 2]
```diff
@@ -466,10 +466,10 @@ impl LogEvent {
     /// or from the `source_type` key set on the "Global Log Schema" (Legacy namespace).
     // TODO: This can eventually return a `&TargetOwnedPath` once Semantic meaning and the
```
Comment (pront) [inline lib/vector-core/src/event/log_event.rs:467]: I will tackle this in a follow-up PR.

File: src/sources/datadog_agent/mod.rs (hunk 1 above)
Comment (StephenWakely) [reply src/sources/datadog_agent/mod.rs:339]: I would prefer to change this to a `String`. This is run at boot time so we shouldn't need to worry about an extra allocation and it just makes for nicer code..

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Looking good. Just changing that `Box::leak` is the only nit I have..

File: src/sources/datadog_agent/mod.rs (hunk 1 above)
Comment (pront) [reply src/sources/datadog_agent/mod.rs]: Makes sense. Changed to `String`.

**Review by StephenWakely — APPROVED** [review]
Looking good!

Just noting [this issue](https://github.com/vectordotdev/vector/issues/17935#issuecomment-1634326118) arose while we were looking through. It breaks in master too, so is not a regression and will be addressed later.

**Review by StephenWakely — APPROVED** [review]

**Pull request merged**